lightnode/chunkgroup/README.md

```diff
@@ -0,0 +1,120 @@
+  The number of chunk groups that a light node participates in is a configuration parameter of the protocol.
+  A light node shuffles each of its chunk groups independently. That is, the light node shuffles its first chunk group
+  assignment at a random time that is independent of the time that it shuffles its second chunk group assignment 
+  (and so on).
+- **Shuffle Epoch**: An shuffle epoch describes the number of times a particular light node has changed chunk groups
+  for a group assignment. At genesis, all light nodes' group assignments are in epoch 0. The epoch for a particular 
+  light node's group assignment is incremented by 1 for each time it is randomly shuffled into a new chunk group. 
+  The length of each epoch is equal to the shuffle period.
+
+
+# Algorithm
+
+The algorithm for determining which chunk group a particular light node is described below. A reference implementation
+of this algorithm can be found in [calculations.go](./calculations.go).
+
+## Determining a node's seed
+
+A light node's seed is an 8 byte value that is randomly assigned to the node at registration time. The seed should
+be generated using on-chain randomness that is difficult for an attacker to predict. The light node's seed
+is public information and stored on-chain.
+
+## `randomInt(seed)`
+
+Define a function `randomInt(seed)` that takes an 8 byte unsigned integer as a seed and returns a pseudo-random 
+8 byte unsigned integer. 
+
+- Copy the 8 byte unsigned integer `seed` into an 8 byte array `seedBytes` in big endian order.
+- Use the `seedBytes` as the input to `keccak256` to generate a 32 byte array called `hashBytes`.
+- Use the first 8 bytes of `hashBytes` to create an 8 byte unsigned integer using big endian order called `result`.
+- Return `result`.
+
+## Function `rotateLeft(x, n)`
+
+Define a function `rotateLeft(x, n)` that shifts the bits in an 8 byte unsigned integer `x` to the left by `n` bits
+and moves the bits that fall off the left side to the right side.
+
+```
+rotateLeft(x, n) := (x << n) | (x >> (64 - n))
+```
+
+## Determining a node's shuffle offset for a particular assignment index
+
+A node's shuffle offset for a particular assignment index is a duration between 0 and the shuffle period at 
+nanosecond granularity. The shuffle offset is determined as follows:
+
+```
+shuffleOffset := randomInt(rotateLeft(nodeSeed, assignmentIndex)) % shufflePeriod_nanoseconds
+```
+
+## Shuffle Epoch Calculation
+
+At the unix epoch, the shuffle epoch for each group assignment is defined as `0`. The epoch increases by `1` at time 
+`shuffleOffset(nodeSeed, assignmentIndex)`, and then by `1` again every `shufflePeriod` nanoseconds.
+
+## Chunk Group Calculation
+
+To determine a node's chunk group, first compute the current epoch for the node, then plug it into the following
+function:
+
+```
+chunkGroup := randomInt(rotateLeft(nodeSeed, assignmentIndex) ^ shuffleEpoch) % numberOfChunks
```

Comment:
I'm unclear on which operation is meant by `^` here. If this is bitwise XOR, then it might be constraining the entropy undesirably.

Reply:
Correct, `^` should be interpreted to mean bitwise XOR.

An alternate strategy might be to just feed the `nodeSeed`, `assignmentIndex`, and `shuffleEpoch` directly into the keccak digest as a byte array.

Reply:
I went ahead and implemented the new strategy that appends the values into a byte array and then feeds the entire byte array into the digest.